cpp/src/explainer/permutation_shap.cu

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuml/explainer/permutation_shap.hpp>
+
+namespace ML {
+namespace Explainer {
+
+template <typename DataT, typename IdxT>
+__global__ void _fused_tile_scatter_pe(DataT* vec, const DataT* bg,
+                                       IdxT nrows_bg, IdxT ncols,
+                                       const DataT* obs, IdxT* idx, IdxT len_bg,
+                                       IdxT sc_size, bool row_major) {
+  // kernel that actually does the scattering as described in the
+  // descriptions of `permutation_dataset` and `shap_main_effect_dataset`
+  IdxT tid = threadIdx.x + blockDim.x * blockIdx.x;
+
+  if (tid < ncols * nrows_bg) {
+    IdxT row, col, start, end;
+
+    if (row_major) {
+      row = tid / ncols;
+      col = tid % ncols;
+      start = (idx[col] + 1) * len_bg;
+      end = start + sc_size * len_bg;
+
+      if ((start <= row && row < end)) {
+        vec[row * ncols + col] = obs[col];
+      } else {
+        vec[row * ncols + col] = bg[(row % len_bg) * ncols + col];
+      }
+
+    } else {
+      col = tid / nrows_bg;
+      row = tid % (len_bg);
```

Comment:
what's the difference between nrows_bg and len_bg? Also, not sure why it's not `row = tid % nrows_bg`?